### DISCUSSION TITLE
ValueError: Cannot get dimension 'nO' for model 'sparse_linear': value unset

### DISCUSSION THREAD

**Discussion by NixBiks** [explosion/spaCy#10196] [Help: Other Questions] [answered]
## How to reproduce the behaviour
I have a pipeline with three components that can't work together for some reason. Currently I can't make a minimal reproducible example right away - I'd need some pointers on how to locate the issue.

The following pipeline

```python
pipeline = ["textcat_multilabel", "sentencizer", "financial_entities"]
```

throws this error (when loading the model)

```
ValueError: Cannot get dimension 'nO' for model 'sparse_linear': value unset
```

Diving deeper into the error message, reveals that specifically it breaks when building `financial_entities` that wraps an extended `EntityRuler`. I.e. when running `ruler.add_patterns`.

The very odd thing is that the pipeline works just fine in these constellations

```python
pipeline = ["textcat_multilabel", "financial_entities"]
pipeline = ["sentencizer", "financial_entities"]
pipeline = ["textcat_multilabel", "sentencizer"]
```

Any ideas how to debug this?

## Your Environment

- **spaCy version:** 3.2.1
- **Platform:** macOS-12.1-arm64-arm-64bit
- **Python version:** 3.8.12

**Comment by NixBiks** [comment]
Okay so I finally got a little wiser on the issue.

My `textcat_multilabel` model was trained with a modified tokenizer using 
```
before_init = {"@callbacks":"customize_tokenizer"}
``` 
in `config.cfg`. 

In `__init__` in my `financial_entities` I have `ruler.add_patterns`. If I instead move that to a method, e.g. `add_patterns` on `financial_entities` and then call that _after_ the pipeline has been loaded. So

```
nlp = spacy.load("./pipeline")
nlp.get_pipe("financial_entities").add_patterns()
```

All in all I'm confused here but it seems to me that I have one tokenizer when `__init__` all my components. And then the modified tokenizer after all `__init__` have been called. Does that sound right?

**Comment by adrianeboyd** [comment, marked as answer]
Because entity ruler phrase patterns run the pipeline (up to the current position in the pipeline), you want to load patterns in `initialize` rather than `__init__` so that the previous components have already been initialized. The entity ruler itself already supports this directly, but you can add whatever additional kwargs to your method that you need for the custom component: https://spacy.io/api/entityruler#initialize

**Reply by NixBiks** [reply]
Great thanks. Quick follow up;

Does that mean I should run `nlp.initialize()` before I can use the pipeline?

```python
nlp = spacy.load("./pipeline")
nlp.initialize()
doc = nlp("Some text...")
```

**Reply by adrianeboyd** [reply]
Probably not. Normally the `initialize` step happens as part of `spacy train` or `spacy assemble`, so if you have something that can be loaded with `spacy.load`, it should have already been initialized. If you call `nlp.initialize` at this point, it will wipe out all the model weights in the statistical models and for some non-trainable components like the entity ruler it would clear the current patterns, too.

**Reply by NixBiks** [reply]
Okay then I don't understand how to do it. I need to make sure that my `initialize` method will be called and thereby adds the patterns when I use `spacy.load`. I assume I won't have to retrain. Currently this is how it looks 

```python
    def initialize(self, get_examples: Callable[[], Iterable[Example]], *, nlp: Optional[Language] = None):
        print("Initializing financial entities")
        self.first_ruler.initialize(get_examples, nlp=nlp, patterns=self.first_rule_patterns)
        self.second_ruler.initialize(get_examples, nlp=nlp, patterns=self.second_rule_patterns)
```

If I understand what you are saying then I need to run either `spacy train` or `spacy assemble` (the latter is unknown to me). I don't want to retrain since I have ~35k documents. If I run `spacy assemble then I get the following error

    ValueError: [E143] Labels for component 'textcat_multilabel' not initialized

I did train it and I also have a `meta.json` with the results so I don't understand that error.

Another approach could be to create a callback which adds the patterns but it seems wrong - I'd assume the component itself should have this logic as with the `initialize` approach.

**Reply by NixBiks** [reply]
I just realized the solution. Put it in `from_disk` instead.

```python
    def from_disk(
        self, path, *, exclude
    ):
        self.first_ruler.add_patterns(self.first_rule_patterns)
        self.second_ruler.add_patterns(self.second_rule_patterns)
```

**Reply by adrianeboyd** [reply]
You'd want `initialize` to take the patterns as kwargs following `nlp` so that they are only loaded from some local path once, and not every time the pipeline is loaded. You don't want to load from any local paths as settings in `__init__` or in settings in the `[components]` block. Then you would need to make sure that your custom component can serialize and reload the entity ruler patterns as part of `to/from_disk/bytes` instead of relying on local paths.

After you have modified the custom component, you can put together the final pipeline with `spacy assemble`, which would use `source` to include already-trained components like a `textcat` and would initialize the remaining components.

If you can restructure the pipeline to use top-level entity ruler components instead, this might be simpler to implement overall since they already handle the serialization for you. If your component doesn't have any other state/settings it needs to be aware of, then you could use `Language.component` instead of `Language.factory`.

**Reply by NixBiks** [reply]
I can't use the original entity ruler unfortunately since I extend it, customize it and wrap a few of them. All my patterns are written in python so the easiest solution is the solution above where I only implement `from_disk`, not `to_disk` nor `initialize`. No need to serialize, but when deserializing the pipeline I add patterns (already loaded in python memory) via `from_disk`. Of course I could also serialize all the patterns to JSON and deserialize from JSON but there is no need really since I write them in python. My custom components with the data hardcoded I guess.

Thanks for your help and please correct me if I'm wrong in my conclusions here.

**Reply by adrianeboyd** [reply]
Without more concrete details about the custom component implementation: I am not really sure.

If you're generating all the patterns in `__init__` without referencing any external data, then you could potentially initialize a phrase matcher from docs created with `nlp.make_doc` rather than `nlp`, but this isn't an option for the entity ruler, which only expects string phrase patterns.
